camayoc/tests/remote/rho/test_scan.py

```diff
@@ -13,35 +13,239 @@
+    rho_profile_add.close()
+    assert rho_profile_add.exitstatus == 0
+
+    reportfile = '{}-report.csv'.format(profile_name)
+    rho_scan = pexpect.spawn(
+        'rho scan --profile {} --reportfile {} --facts all'
+        .format(profile_name, reportfile),
+        timeout=300,
+    )
+    input_vault_password(rho_scan)
+    rho_scan.logfile = BytesIO()
+    assert rho_scan.expect(pexpect.EOF) == 0
+    logfile = rho_scan.logfile.getvalue().decode('utf-8')
+    rho_scan.logfile.close()
+    rho_scan.close()
+    assert rho_scan.exitstatus == 0, logfile
+    assert os.path.isfile(reportfile)
+    return reportfile
+
+
+@pytest.fixture(scope='module', autouse=True)
+def scan_machines(vcenter_client, isolated_filesystem):
+    """Scan all machines caching the results."""
+    config = get_config()
+    auths = config['rho']['auths']
+    inventory = {
+        machine['hostname']: machine for machine in config['inventory']
+        if machine['hypervisor'] == 'vcenter'
+    }
+    if not auths or not inventory:
+        raise ValueError(
+            'Make sure to have auths and inventory on the configuration file')
+    hostnames = [machine['hostname'] for machine in inventory.values()]
+    host_folder = vcenter_client.content.rootFolder \
+        .childEntity[VCENTER_DATA_CENTER].hostFolder
+    host = host_folder.childEntity[VCENTER_CLUSTER].host[VCENTER_HOST]
+    vms = [
+        vm for vm in host.vm
+        if vm.name in hostnames
+    ]
+    for auth in auths:
+        if auth.get('sshkeyfile'):
+            auth_add({
+                'name': auth['name'],
+                'username': auth['username'],
+                'sshkeyfile': auth['sshkeyfile'],
+            })
+        elif auth.get('password'):
+            auth_add(
+                {
+                    'name': auth['name'],
+                    'username': auth['username'],
+                    'password': None,
+                },
+                [(
+                    CONNECTION_PASSWORD_INPUT,
+                    auth['password']
+                )],
+            )
+    chunk_size = 5
+    chunks = [vms[i:i+chunk_size] for i in range(0, len(vms), chunk_size)]
+    for chunk in chunks:
+        machines_to_wait = []
+        for vm in chunk:
+            if vm.runtime.powerState == 'poweredOff':
+                vm.PowerOnVM_Task()
+                machines_to_wait.append(inventory[vm.name])
+        wait_until_live(machines_to_wait)
+        for vm in chunk:
+            if vm.runtime.powerState == 'poweredOn':
+                machine = inventory[vm.name]
+                for auth in auths:
+                    profile_name = machine['hostname'] + '-' + auth['name']
+                    try:
+                        result = scan_machine(machine, auth, profile_name)
+                    except (AssertionError, pexpect.exceptions.EOF) as err:
+                        result = None
+                    SCAN_RESULTS[profile_name] = result
```

Comment:
To check my understanding, I see that we will if know the scan failed if the SCAN_RESULTS[profile_name] is None

Reply:
but the scanning of all the machines will proceed because it is guarded by this try block

Reply:
Yeah, you've got it.

To provide more context: I decided to have permutations based on auth and the machines available which is generating 4 items per machine. The fact result analysis will be done in a way that all facts that don't have expected values will be printed into the failure message.  This way it should be easy to pin point what is the machine and user raising issues.

Also since I got that people are more interested on knowing if rho is able to collect some set of facts depending on the user privilege and the products installed, I went into some implementations and found that this one should provide the information they are interested and also allow us to check that more quickly.

All that said, let me know if you agree with it. Or have a different point of view.

Reply:
Sounds good to me, I appreciate you addressing the root challenge of figuring out what type of results will be most useful for us to actually diagnose the problem.